### PR TITLE
Consider inheritance in type checking for tensors

### PR DESCRIPTION
**Pull request by daemyung**
# What does this PR do?

Add an additional check to bypass type assertion when both tensors are torch.Tensor instances.

# Why this PR is needed.

I have defined a class which inherit torch.Tensor to keep additional information. Since it inherits from torch.Tensor, it should be acceptable to treat it as torch.Tensor. However the current code requires an exact type match, which doesn't handle my case. I believe this strict type checking does not satisfy to the priciple of OOP.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr and @SunMarc

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
LGTM ! Could you do `make style` to fix the quality issue ?

**Comment by daemyung** [comment]
@SunMarc Thanks. I have fixed the quality issue. Let's wait CI is passed.

**Comment by daemyung** [comment]
@SunMarc CI is passed. Please merge it. Thanks.

File: src/transformers/trainer_pt_utils.py [hunk 1]
```diff
@@ -131,9 +131,10 @@ def nested_concat(tensors, new_tensors, padding_index=-100):
     Concat the `new_tensors` to `tensors` on the first dim and pad them on the second if needed. Works for tensors or
     nested list/tuples/dict of tensors.
     """
-    assert type(tensors) == type(
-        new_tensors
-    ), f"Expected `tensors` and `new_tensors` to have the same type but found {type(tensors)} and {type(new_tensors)}."
+    if not (isinstance(tensors, torch.Tensor) and isinstance(new_tensors, torch.Tensor)):
```
Comment (ArthurZucker) [inline src/transformers/trainer_pt_utils.py:134]: We could simply test that both are instance of `torch.Tensor`. If you inherit from `torch.Tensor` is instance will be true

**Review by ArthurZucker — APPROVED** [review]
Thanks for fixing

File: src/transformers/trainer_pt_utils.py (hunk 1 above)
Comment (daemyung) [reply src/transformers/trainer_pt_utils.py:134]: @ArthurZucker Sorry I can't get your point. Could you let me know how can we simplify more?

**Comment by daemyung** [comment]
@muellerzr and @SunMarc Could you merge this PR?

**Review by muellerzr — APPROVED** [review]
Thanks! LG2M as well. cc @ArthurZucker I think we can merge this no?

**Comment by ArthurZucker** [comment]
Yes! No worries

**Pull request merged**